target/linux/mvebu/cortexa9/base-files/etc/init.d/fancontrol

```diff
@@ -0,0 +1,93 @@
+#!/bin/sh /etc/rc.common
+
+# control pwm fans with basic 10-point fan speed increments
+
+START=98
+
+start() {
+	case $(board_name) in
+	kobol,helios4)
+		#hwmon0 -> ../../devices/platform/soc/soc:internal-regs/f1072004.mdio/mdio_bus/f1072004.mdio-mii/f1072004.mdio-mii:00/hwmon/hwmon0
+		#hwmon1 -> ../../devices/platform/j10-pwm/hwmon/hwmon1
+		#hwmon2 -> ../../devices/platform/j17-pwm/hwmon/hwmon2
+		#hwmon3 -> ../../devices/virtual/thermal/thermal_zone0/hwmon3
+		cpu_temp_path="/sys/class/hwmon/hwmon3/temp1_input" # temperature of CPU as sensed by internal sensor, millicelsius
+		eth_temp_path="/sys/class/hwmon/hwmon3/temp1_input" # temperature of ethernent controller over MDIO bus, millicelsius, not used in the script
+		fan1_pwm_path="/sys/class/hwmon/hwmon1/pwm1" #fan connected on j10 header on board
+		fan2_pwm_path="/sys/class/hwmon/hwmon2/pwm1" #fan connected on j17 header on board
+				
+		pwm_min="20" #lowest value for fan speed, with pwm lower than this value the fan is off
+		
+		cpu_temp_min="40000" #lowest value, if CPU is cooler than this fan is off
+		cpu_temp_max="85000" #max value, if CPU is hotter the fan is at max speed
+		;;
+	esac
+
+	#calculating 10 temperature steps
+	temp_step=$(( (cpu_temp_max - cpu_temp_min) / 10 ))
+	
+	target_temp_1=$(( cpu_temp_min + ( temp_step * 1 ) ))
+	target_temp_2=$(( cpu_temp_min + ( temp_step * 2 ) ))
+	target_temp_3=$(( cpu_temp_min + ( temp_step * 3 ) ))
+	target_temp_4=$(( cpu_temp_min + ( temp_step * 4 ) ))
+	target_temp_5=$(( cpu_temp_min + ( temp_step * 5 ) ))
+	target_temp_6=$(( cpu_temp_min + ( temp_step * 6 ) ))
+	target_temp_7=$(( cpu_temp_min + ( temp_step * 7 ) ))
+	target_temp_8=$(( cpu_temp_min + ( temp_step * 8 ) ))
+	target_temp_9=$(( cpu_temp_min + ( temp_step * 9 ) ))
+	target_temp_10=$cpu_temp_max
```

Comment:
I wonder whether all of that could be done much easier with a for loop (pseudo-code):

for i in 0:1:10
   if realtemp > $(( cpu_temp_min + ( temp_step * i ) ))
      continue
   fi
   pwm_value = $(( pwm_min + ( ( 255 - pwm_min ) / 10 * i ) ))
   break
fi

Reply:
the idea was to precompute the values so it would not need to do the same calculation over and over again every time it needs to adjust the temp.

Reply:
That's what I thought, but is it really worth it?

Reply:
I'd say so.

Reply:
For a script of this size, I doubt it. But that's my habit.

Executing it all once (removing the while loop at the end) aka it's calculating all values on the Helios4 device (in Armbian, the current OS it is running) I get 
time bash -c ./test-script 
./test-script: 90: ./test-script: cannot create /sys/class/hwmon/hwmon1/pwm1: Permission denied
./test-script: 91: ./test-script: cannot create /sys/class/hwmon/hwmon2/pwm1: Permission denied

real	0m0.012s
user	0m0.001s
sys	0m0.012s


Still, what does "easier" mean for your proposed change? 

Should be something like this in shell script (not tested)
```
	for i in {1..10} ; do
	if [ $cpu_temp -le $(( cpu_temp_min + ( ( (cpu_temp_max - cpu_temp_min) / 10 ) * i ) )) ]; then
	    continue
	else
	    pwm_value=$(( pwm_min + ( ( 255 - pwm_min ) / 10 * i ) ))
	    break
	fi
```

Imho it's harder to understand at a glance what is going on, and these devices have ample flash space or a SD card, so even if it is longer it does not matter much.

Reply:
I run this thing on Armbian. It's quite slow (I probably have no business running docker on this). Any speed improvement is welcome.

Reply:
> I run this thing on Armbian. It's quite slow (I probably have no business running docker on this). Any speed improvement is welcome.

You mean Armbian on another device, or you have a Helios 4 too?

Because my benchmark above with Helios 4 running Armbian shows running a script with all calculations only needs 0.012 seconds. The device has a pretty decent CPU.

Imho if you have a Armbian device and it runs bad you need to check RAM consumption and maybe disable zram. I have seen enough times that when I use zram on anything (not just on Armbian) and the device does not have enough free memory the devices get stuck in a loop of "compressing and moving to zram" and "decompressing and moving to ram" and this kills any performance.

Reply:
I have a Helios 4. I also have a Turris zombis. Both are the same generation CPU at the same clockspeed. Nextcloud on TurrisOS performs faster than on Armbian.